lib/spack/llnl/util/lock.py

```diff
@@ -267,24 +340,65 @@ def acquire_write(self, timeout=None):
         """
         timeout = timeout or self.default_timeout
 
+        locktype = 'WRITE LOCK'
         if self._writes == 0:
-            self._debug(
-                'WRITE LOCK: {0.path}[{0._start}:{0._length}] [Acquiring]'
-                .format(self))
+            self._log_acquiring(locktype)
             # can raise LockError.
             wait_time, nattempts = self._lock(fcntl.LOCK_EX, timeout=timeout)
-            self._acquired_debug('WRITE LOCK', wait_time, nattempts)
             self._writes += 1
+            self._log_acquired(locktype, wait_time, nattempts)
 
             # return True only if we weren't nested in a read lock.
             # TODO: we may need to return two values: whether we got
             # the write lock, and whether this is acquiring a read OR
             # write lock for the first time. Now it returns the latter.
             return self._reads == 0
         else:
+            # Nothing is done with the provided lock timeout so log a warning
+            # if it is different from the one being used.
+            self._log_timeout_warning(timeout)
+
             self._writes += 1
             return False
 
+    def downgrade_write(self, timeout=None):
+        """
+        Downgrade from an exclusive write lock to a shared read.
+
+        Raises:
+            LockDowngradeError: if this is an attempt at a nested transaction
+        """
+        timeout = timeout or self.default_timeout
+
+        if self._writes == 1 and self._reads == 0:
+            self._log_downgrading()
+            # can raise LockError.
+            wait_time, nattempts = self._lock(fcntl.LOCK_SH, timeout=timeout)
+            self._reads = 1
+            self._writes = 0
+            self._log_downgraded(wait_time, nattempts)
+        else:
+            raise LockDowngradeError(self.path)
+
+    def upgrade_read(self, timeout=None):
```

Comment:
Same question here -- can we change this to `read_to_write` or `upgrade_read_to_write`?

Reply:
Sure.  Picked the latter.